pkg/combination/combination.go

```diff
@@ -51,39 +58,85 @@ func WithArgs(args map[string][]string) StreamOption {
 // only occurs on the first call to Next or All. By using this, the Stream
 // will solve all possible combinations of its args which could take a lot
 // of computation given a large enough input.
-func WithSolveAhead() StreamOption {
+func WithSolveAhead(boolVar ...bool) StreamOption {
+	boolVarValue := false
+	if len(boolVar) > 0 {
+		boolVarValue = boolVar[0]
+	}
 	return func(cs *stream) {
-		cs.solveAhead = true
+		cs.solveAhead = boolVarValue
 	}
 }
 
-// Next gets the next combination from the stream.
-// TODO: Currently this does not solve each combination iteratively and will
-//       need to do so in the future to ensure an optimal use of memory. This
-//       is currently more of a stub to allow consumer packages to maintain its
-// 		 interface.
-func (cs *stream) Next(ctx context.Context) (map[string]string, error) {
-	if cs.solveAhead && !cs.solved {
-		if err := cs.solve(); err != nil {
-			return nil, err
-		}
+// nextIterativeCombination() gets the next combination from the stream.
+// By using this, the stream will solve each combination
+// iteratively.
+func (cs *stream) nextIterativeCombination() (map[string]string, error) {
+	if cs.solved {
+		return nil, nil
 	}
+	// Check to see if anymore combinations exist
 
-	if len(cs.combinations) == 0 {
-		if cs.solved {
-			return nil, nil
-		}
-		return nil, ErrCombinationsNotSolved
+	// Edge case: No parameterListFromArgs
+	if len(cs.parameterListFromArgs) == 0 {
+		cs.solved = true
+		return nil, ErrNoArgsSet
 	}
 
-	next := cs.combinations[0]
-	cs.combinations = cs.combinations[1:]
+	// comboList is a variable that holds a list of combinations to be returned.
+	comboList := map[string]string{}
+
+	// Generate the list of combinations based off current positions
+	for x := 0; x < len(cs.parameterListFromArgs); x++ {
+		combo := cs.args[cs.parameterListFromArgs[x]][cs.positionsMapInArgs[x]]
+		key := cs.parameterListFromArgs[x]
+		comboList[key] = combo
+	}
 
-	return next, nil
+	// Iterates through position map in reverse
+	// looking for the first updatable value then breaks loop.
+	// Otherwise, resets values to zero, we know to update last parameter based off i.
+	var i int
+	for i = len(cs.positionsMapInArgs) - 1; i > cs.nextParameterToUpdate; i-- {
+		if cs.positionsMapInArgs[i]+1 < len(cs.args[cs.parameterListFromArgs[i]]) {
+			cs.positionsMapInArgs[i]++
+			break
+		} else {
+			cs.positionsMapInArgs[i] = 0
+		}
```

Comment:
nit/not blocking: you can reduce the amount of nesting by removing the if/else conditional structure due to the `break` keyword being used:

```go
if cs.positionsMapInArgs[i]+1 < len(cs.args[cs.parameterListFromArgs[i]]) {
	cs.positionsMapInArgs[i]++
	break
}
cs.positionsMapInArgs[i] = 0
```